modules/table/src/test/java/org/apache/ignite/internal/table/distributed/IndexGcTest.java

```diff
@@ -0,0 +1,340 @@
+                hashInnerStorage,
+                USER_INDEX_BINARY_TUPLE_CONVERTER::toTuple
+        );
+
+        storage = new TestMvPartitionStorage(1);
+
+        storageUpdateHandler = new StorageUpdateHandler(1, new TestPartitionDataStorage(storage),
+                () -> Map.of(
+                        pkIndexId, pkStorage,
+                        sortedIndexId, sortedIndexStorage,
+                        hashIndexId, hashIndexStorage
+                )
+        );
+    }
+
+    @Test
+    void testRemoveStaleEntryWithSameIndex() {
+        UUID rowUuid = UUID.randomUUID();
+        RowId rowId = new RowId(1, rowUuid);
+
+        var key = new TestKey(1, "foo");
+        var value = new TestValue(2, "bar");
+        BinaryRow tableRow = binaryRow(key, value);
+
+        addWrite(storageUpdateHandler, rowUuid, tableRow);
+        commitWrite(rowId);
+
+        addWrite(storageUpdateHandler, rowUuid, tableRow);
+        commitWrite(rowId);
+
+        assertEquals(2, getRowVersions(rowId).size());
+        assertThat(pkInnerStorage.allRowsIds(), contains(rowId));
+        assertThat(sortedInnerStorage.allRowsIds(), contains(rowId));
+        assertThat(hashInnerStorage.allRowsIds(), contains(rowId));
+
+        assertTrue(storageUpdateHandler.vacuum(CLOCK.now()));
+
+        assertEquals(1, getRowVersions(rowId).size());
+        // Newer entry has the same index value, so it should not be removed.
+        assertTrue(inIndex(tableRow));
+    }
+
+    @Test
+    void testRemoveStaleEntriesWithDifferentIndexes() {
+        UUID rowUuid = UUID.randomUUID();
+        RowId rowId = new RowId(1, rowUuid);
+
+        var key1 = new TestKey(1, "foo");
+        var value1 = new TestValue(2, "bar");
+        BinaryRow tableRow1 = binaryRow(key1, value1);
+
+        var key2 = new TestKey(1, "foo");
+        var value2 = new TestValue(5, "baz");
+        BinaryRow tableRow2 = binaryRow(key2, value2);
+
+        addWrite(storageUpdateHandler, rowUuid, tableRow1);
+        commitWrite(rowId);
+
+        addWrite(storageUpdateHandler, rowUuid, tableRow1);
+        commitWrite(rowId);
+
+        addWrite(storageUpdateHandler, rowUuid, tableRow2);
+        commitWrite(rowId);
+
+        assertEquals(3, getRowVersions(rowId).size());
+        assertThat(pkInnerStorage.allRowsIds(), contains(rowId));
+        assertThat(sortedInnerStorage.allRowsIds(), contains(rowId));
+        assertThat(hashInnerStorage.allRowsIds(), contains(rowId));
+
+        HybridTimestamp afterCommits = CLOCK.now();
+
+        assertTrue(storageUpdateHandler.vacuum(afterCommits));
+        assertTrue(storageUpdateHandler.vacuum(afterCommits));
+        assertFalse(storageUpdateHandler.vacuum(afterCommits));
+
+        assertEquals(1, getRowVersions(rowId).size());
+        // Older entries have different indexes, should be removed.
+        assertFalse(inIndex(tableRow1));
+        assertTrue(inIndex(tableRow2));
+    }
+
+    @Test
+    void testRemoveTombstones() {
```

Comment:
I think you need to test the following scenarios:
- null row row
- row null row
- row row null
- current implementation